Hashable type class

fehrenbach opened the pull request:
See #188.

fehrenbach committed:
Add Hashable class

fehrenbach committed:
Improve Hashable Boolean

fehrenbach committed:
Spaces and whatnot

kritzcreek commented:
This looks like a great start! I'm assuming you didn't just come up with these hashing functions and took them from somewhere else? (The Haskell implementation? Some paper?)

I think it would be good to have a few comments referencing where these hashing functions are coming from, to have some form of rationale for them.

fehrenbach commented:
The one for Number is used by immutable.js, among others.
Haskell uses identity for Int and fromEnum for "small" types.
The String and Array algorithm here is the same as Java, I think.

I can add comments. It'd be even better if someone did some actual research on the matter. We are in the fortunate position that Eq is homogeneous and there is no subtyping, so changing the algorithm is not a breaking change and can be done for any type individually.

fehrenbach committed:
Add comments about hash functions

and fix the record hash function in the process

fehrenbach committed:
Hash code is an indexed newtype

fehrenbach commented:
This now uses the indexed newtype as discussed in #188.

LiamGoodacre commented on src/Data/Hashable.purs:
```diff
@@ -0,0 +1,118 @@
+-- | are never equal. The reverse is not necessarily true.
+-- |
+-- | Hash values produced by `hash` must not be relied upon to be
+-- | stable accross multiple executions of a program and should not be
```
Typo: `accross` -> `across`

fehrenbach replied:
Thanks!

fehrenbach committed:
Fix spelling

fehrenbach committed:
Make length contribute to array hash

Hash values of 0 are not uncommon, e.g., 0, false, Nothing, {}, ...
We do not want to hash [true], [false, true], [false, false, true], ... to the same value.

fehrenbach commented:
Thanks to @rightfold for pointing out the array hash problem.

JordanMartinez commented:
@fehrenbach Should this PR be closed in favor of #198?

fehrenbach commented:
Yeah, probably. I should get back to this at some point...
